### PR TITLE
feat(perms): add permission checking for moving resources

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem
we need to add permissions checking for moving items between folders. 

Closes [insert issue #]

## Solution
add initial permissions check. this allows editors to freely move **non-root** items between **non-root** resources. we allow admins to freely move items between folders.

## Screenshots

https://github.com/user-attachments/assets/666f286a-efbe-4be3-b3d1-0fbd36f5d6e6

### CODE REVIEW

**Comment by seaerchin** [comment]
note that this allows admins to move **root level folders**

**Comment by seaerchin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/712?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#714** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/714?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#712** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/712?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#693** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/693?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#683** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/683?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#681** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/681?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @seaerchin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/studio/src/server/modules/resource/resource.router.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.router.ts:1]: from karrui: actually we should have a function that does a traversal to root next time, we can write this function now but return `null`

File: apps/studio/src/server/modules/resource/resource.router.ts [hunk 1]
```diff
@@ -17,6 +17,46 @@ import {
...
+  from: string
+  to: string | null
+}) => {
+  // TODO: this is using site wide permissions for now
+  // we should fetch the oldest `parent` of this resource eventually.
+  // Putting this in here first because eventually we'll have to lookup both
+  // even though for now they are the same thing
+  const permsFrom = await definePermissionsFor({ ...rest, resourceId: null })
+  const permsTo = await definePermissionsFor({ ...rest, resourceId: null })
+
```
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.router.ts:51]: we should pass to `from, to` then call the traversal methods

File: apps/studio/src/server/modules/resource/resource.router.ts
Comment (seaerchin) [reply apps/studio/src/server/modules/resource/resource.router.ts:1]: we should just remove the `move` permission and use `update`; users can move if they have `delete` perms on the moved resource and `create` perms on the destination

File: apps/studio/src/server/modules/resource/resource.router.ts [hunk 2]
```diff
@@ -117,52 +157,71 @@ export const resourceRouter = router({
...
+          siteId,
         })
-    }),
+        if (!isValid) {
+          throw new TRPCError({
+            code: "FORBIDDEN",
+            message:
+              "Please ensure that you have sufficient permissions to move the specified resource",
+          })
+        }
```
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.router.ts:177]: should just throw in the `validateUserPermission` function

File: apps/studio/src/server/modules/resource/resource.router.ts [hunk 3]
```diff
@@ -146,21 +186,82 @@ export const resourceRouter = router({
...
+          .executeTakeFirst()
+      })
+
+      if (!isValid) {
+        throw new TRPCError({
+          code: "FORBIDDEN",
+          message:
+            "Please ensure that you have the required permissions to perform a move!",
+        })
+      }
```
Comment (dcshzj) [inline apps/studio/src/server/modules/resource/resource.router.ts:256]: nit: can't really tell from the code review interface but this (and below) appears to be unreachable code?

File: apps/studio/src/server/modules/resource/resource.router.ts [hunk 4]
```diff
@@ -146,21 +186,82 @@ export const resourceRouter = router({
   move: protectedProcedure
     .input(moveSchema)
     .mutation(async ({ input: { movedResourceId, destinationResourceId } }) => {
+      const isValid = await validateUserPermissions({
+        from: movedResourceId,
+        to: destinationResourceId,
+        userId: ctx.user.id,
```
Comment (dcshzj) [inline apps/studio/src/server/modules/resource/resource.router.ts:192]: issue: `ctx` seems to be missing here

**Review by dcshzj — APPROVED** [review]

**Comment by seaerchin** [comment]
### Merge activity

* **Oct 22, 1:43 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/isomer/712).

**Pull request merged**